### PR TITLE
feat: min max leverage

### PR DESCRIPTION
**Pull request by avniculae**
# Min Max Leverage

## Description
The initiative involves the following:
- raising a Reminder (warning) if user uses low leverage
- computing the highest leverage possible and include it in the leverage buttons as:
      - MAX_LEVERAGE / 4, MAX_LEVERAGE / 2, MAX_LEVERAGE
- exclude fees when computing the leverage
- fix (some) bugs between the interaction of: notional, margin and leverage

## Amplify hosted environment link

https://feat-minmaxleverage.d2f2fhja7tsifl.amplifyapp.com/

## Notion link

[Notion link](https://www.notion.so/voltz/Min-Max-Leverage-7d91410b2a9e4eb4a268f15cec85c353)

### CODE REVIEW

File: src/components/interface/SwapInfo/SwapInfo.tsx [hunk 1]
```diff
@@ -92,11 +92,16 @@ export const SwapInfo: React.FunctionComponent<SwapInfoProps> = ({
...
       {(mode === SwapFormModes.NEW_POSITION || mode === SwapFormModes.FIX_BORROW) &&
-        !isUndefined(warningText) && (
+        !isUndefined(warningBoxes) && (
           <>
-            <Box sx={bottomSpacing}>
-              <WarningBox warningText={warningText} />
-            </Box>
+            {warningBoxes.map((warningBox) => {
+              return (
+                // eslint-disable-next-line react/jsx-key
```
Comment (ssbarbee) [inline src/components/interface/SwapInfo/SwapInfo.tsx:99]: there is a reason why we need to use keys for Lists and why this is enforced via Lint. Please remove the eslint-disable-next-line react/jsx-key and pass a key to the Box component.

Documentation: https://reactjs.org/docs/lists-and-keys.html

File: src/contexts/SwapFormContext/SwapFormContext.tsx [hunk 2]
```diff
@@ -413,15 +419,19 @@ export const SwapFormProvider: React.FunctionComponent<SwapFormProviderProps> =
...
+        swapInfo.result?.availableNotional !== undefined &&
+        !isUndefined(swapInfo.result?.fee)
       ) {
         if (!touched.current.includes('margin')) {
           touched.current.push('margin');
         }
         const minNotional = Math.min(notional, swapInfo.result?.availableNotional);
-        const formatted = formatNumber(minNotional / newLeverage, 0, 4);
+        const formatted = formatNumber(minNotional / newLeverage + swapInfo.result?.fee, 0, 4);
         const newMargin = stringToBigFloat(formatted);
```
Comment (ssbarbee) [inline src/contexts/SwapFormContext/SwapFormContext.tsx:430]: what if swapInfo.result?.fee is undefined? you don't have this in the if above

File: src/components/interface/SwapInfo/SwapInfo.tsx (hunk 1 above)
Comment (avniculae) [reply src/components/interface/SwapInfo/SwapInfo.tsx:99]: adressed

File: src/utilities/number/index.ts [hunk 3]
```diff
@@ -113,3 +113,30 @@ export const compactFormat = (value: number): string => {
   });
   return formatter.format(value);
 };
+
+/**
+ * It takes a floating point leverage and formats it to an rounded integer
+ * @param leverage The leverage number to format (can be float)
+ * @returns The formatted leverage (only integer)
+ */
+export const formatLeverage = (leverage: number): number => {
```
[the remark this answers is not in the document]
Comment (avniculae) [reply src/utilities/number/index.ts:122]: addressed

**Review by ssbarbee — APPROVED** [review]

**Pull request merged**